### ISSUE TITLE
radpred returns NA for extreme ranks

### ISSUE THREAD

**Issue by piklprado** [piklprado/sads#24]
As radpred uses approxfun, sometimes abundances for extreme ranks (usually the first) returns NA:

```r
teste <- c(1,38,12,2,1,1,6,19,12,5,6,37,1,20,2,5,2,1,21)
teste.bs <- fitsad(teste, "bs")
radpred(teste.bs)
Object of class "rad"
   rank abund
1     1    NA
2     2    33
3     3    25
4     4    20
5     5    17
6     6    14
7     7    13
8     8    11
9     9     9
10   10     8
11   11     7
12   12     6
13   13     5
14   14     4
15   15     4
16   16     3
17   17     2
18   18     1
19   19     1
```

**Comment by andrechalom** [comment]
I may be wrong here, but I think this can be fixed by setting "rule=2" in the approxfun call. From the manual:

**rule**: an integer describing how interpolation is to take place outside the interval [‘min(x)’, ‘max(x)’].  If ‘rule’ is ‘1’ then ‘NA’s are returned for such points and if it is ‘2’, the value at the closest data extreme is used.

The following code has been modified from radpred to illustrate the rule=2 behavior:
```R
teste <- c(1,38,12,2,1,1,6,19,12,5,6,37,1,20,2,5,2,1,21)
teste.bs <- fitsad(teste, "bs")
# Old code
print(radpred(teste.bs)[1,])
# New code
S <- length(teste.bs@data$x)
y <- 1:sum(teste.bs@data$x)
X <- do.call(pbs, c(list(q = y, lower.tail = F), as.list(bbmle::coef(teste.bs))))
f1 <- approxfun(x=c(1, X), y=c(0, y), method="constant", rule=2)
ab <- f1(ppoints(S))
print(data.frame(rank=1:S, abund=ab)[1,])
```
This code prints:
```R
  rank abund
1    1    NA
  rank abund
1    1   181
```

**Comment by andrechalom** [comment]
Also on this topic, if the result from the approxfun will be used only once to generate the "ab" object, why are we using approxfun instead of approx?
```R
ab <- approx(x=c(1, X), y=c(0, y), xout=ppoints(S), method="constant", rule=2)$y
```

**Comment by piklprado** [comment]
Still, `rule=2` only replaces the missing values with the largest value, right? 
So it is an approximation with variable bias. We'd better look for a solution that returns the exact expected value for all ranks. As rank-abundance plots show the inverse function of a [p] function with lower.tail=FALSE. Using `rule=2` we'll replace NA's in the first ranks with the inverse of [p](max(y)) where y is the vector of observed abundances.
As inv[p] = [q] this could be solved using quantile functions as is already done for continuous distributions. But I gave up to do that with the  discrete distributions because old  [q] code was too slow. Maybe the improvements in the quantile  functions (issue #35)

**Comment by andrechalom** [comment]
Yes, you're right. 

I've changed the code to use the [q] function for all distributions, but it's still quite slow for some distributions on the larger datasets. `radpred(fitpower(okland))` takes some 3-4 seconds, but `radpred(fitpower(moths))` is hanging for full 5 minutes (the code with approx takes _only_ 15 secs)

So, I wrote a "compromise" version that uses initially uses approx([p]...), then checks if there is any NA. In that case, it runs [q] explicitly. This version takes... 5 minutes to run the moths example above, of which 4.5 minutes are spent calculating the [q] of the lowest ppoint. I'll try to see if I can make the [q] code run faster...

However, it should be noted that the extreme ranks being predicted are **definitely** extreme:
```R
> radpred(fitpower(moths))[1,2]
Note: extreme values generated by radpred. Calculations will take a while...
[1] 118984350
> max(moths)
[1] 2349
```

So, I don't see any solution here, other than notifying the user trying to fit these extreme sets, and working on a way to make [q] code run faster...

On to another notes: 
1. The code from radpred had loads of duplicated and unnecessary lines. I've condensed all the methods that receive fitsad, fitrad, or abundance objects to call two "base" methods which receive sad/rad, coef, S and N.
2. The handling of dots in the radpred methods was extremely inconsistent. Some signatures passed the dots to the [q] function, some silently ignored them, and some did not accept any argument except those listed. I'm removing the dots from all signatures to keep everything standardized, as I really don't see any meaningful way to use the dots.
3. The radpred from "bs" fits is generating an error, however, but that's due to issue #42.
